tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,584 @@
+ * limitations under the License.
+ */
+
+package com.pingcap.tikv;
+
+import com.google.protobuf.ByteString;
+import com.pingcap.tikv.exception.GrpcException;
+import com.pingcap.tikv.exception.TiBatchWriteException;
+import com.pingcap.tikv.region.RegionManager;
+import com.pingcap.tikv.region.TiRegion;
+import com.pingcap.tikv.txn.TxnKVClient;
+import com.pingcap.tikv.txn.type.BatchKeys;
+import com.pingcap.tikv.txn.type.ClientRPCResult;
+import com.pingcap.tikv.txn.type.GroupKeyResult;
+import com.pingcap.tikv.util.BackOffFunction;
+import com.pingcap.tikv.util.BackOffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.tikv.kvproto.Kvrpcpb;
+
+public class TiBatchWrite2PCClient {
+
+  public static class ByteWrapper {
+    private byte[] bytes;
+
+    public ByteWrapper(byte[] bytes) {
+      this.bytes = bytes;
+    }
+
+    public byte[] getBytes() {
+      return this.bytes;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      ByteWrapper that = (ByteWrapper) o;
+
+      return Arrays.equals(bytes, that.bytes);
+    }
+
+    @Override
+    public int hashCode() {
+      return Arrays.hashCode(bytes);
+    }
+  }
+
+  public static class BytePairWrapper {
+    private byte[] key;
+    private byte[] value;
+
+    public BytePairWrapper(byte[] key, byte[] value) {
+      this.key = key;
+      this.value = value;
+    }
+
+    public byte[] getKey() {
+      return key;
+    }
+
+    public byte[] getValue() {
+      return value;
+    }
+  }
+
+  /** buffer spark rdd iterator data into memory */
+  private static final int WRITE_BUFFER_SIZE = 1024 * 1024;
+
+  /**
+   * TiKV recommends each RPC packet should be less than ~1MB. We keep each packet's Key+Value size
+   * below 16KB.
+   */
+  private static final int TXN_COMMIT_BATCH_SIZE = 16 * 1024;
+
+  /** unit is second */
+  private static final long DEFAULT_BATCH_WRITE_LOCK_TTL = 3000;
+
+  private static final Logger LOG = LoggerFactory.getLogger(TiBatchWrite2PCClient.class);
+
+  private TxnKVClient kvClient;
+  private RegionManager regionManager;
+
+  /** start timestamp of transaction which get from PD */
+  private final long startTs;
+
+  public TiBatchWrite2PCClient(TxnKVClient kvClient, long startTime) {
+    this.kvClient = kvClient;
+    this.regionManager = kvClient.getRegionManager();
+    this.startTs = startTime;
+  }
+
+  /**
+   * 2pc - prewrite primary key
+   *
+   * @param backOffer
+   * @param primaryKey
+   * @param value
+   * @return
+   */
+  public void prewritePrimaryKey(BackOffer backOffer, byte[] primaryKey, byte[] value)
```

Comment:
In RegionStoreClient, Key and Value are `ByteString`. It is better to replace `byte[]` with  `ByteString`

Reply:
I think both is ok.

Reply:
ByteString should be better